### ISSUE TITLE
Electron - Better support for devtools extensions

### ISSUE THREAD

**Issue by zcbenz** [github/gsoc#98]
**Summary:**

Electron has basic support for devtools extensions, but still lacks support for popular devtools extensions like React DevTools Extension.

**Expected Outcome:**
- Implement basic `chrome.*` APIs to make React DevTools Extensions work.
- Have a complete test case for DevTools Extensions APIs.
- Provide a built-in GUI interface to manage DevTools Extensions.

**Skills**: C++, JavaScript, Chrome Extensions

**Relevant issues:**
- https://github.com/atom/electron/issues/915

Mentors: @zcbenz

**Comment by royxue** [comment]
@zcbenz Hi, Cheng, I find this issue is interesting, I would like to do it.

**Comment by faurehu** [comment]
@zcbenz I am also interested, will apply.

**Comment by chanlvh** [comment]
I'm also interested :) Will apply soon

**Comment by kishoredbn** [comment]
@zcbenz Hi Cheng, This seems something feasible. Can you assign any task for me as for now so that I can work on that and analyse how comfortable I'll be going forward with this project, and to be considered for this project.

**Comment by nirlanka** [comment]
@zcbenz, I have experience in working with and developing devtool extensions in Chrome. This seems like a really interesting and useful feature in making Electron more powerful in handling development on every level including hacking itself and I'd love to work on this.

**Comment by bridiver** [comment]
I am currently working on implementing the chrome.\* APIs and have made significant progress. I'll have updates for electron and libhchromiumcontent soon. Our minimum api requirements are based around other extensions, but it probably covers a lot of what is necessary for devtools

**Comment by bridiver** [comment]
does anyone have a list of the APIs that are required for this ticket? I have the "common" APIs like webRequest and runtime working and partial support for browser APIs like windows, tabs and browserAction. I also implemented a chrome.ipc API that gives content scripts and background pages access to ipcRenderer. Based on #177 it sounds like only content scripts are needed and those are working with complete support for all the APIs that are available in that context.

**Comment by zcbenz** [comment]
@bridiver This is awesome! We don't really have a list of APIs required, for react-devtools only, I think `chrome.runtime.connect` is required:
https://github.com/facebook/react-devtools/search?p=2&q=%22chrome.%22&type=Code&utf8=✓

Also are you going to attend Google Summer of Code? 😸

**Comment by bridiver** [comment]
@zcbenz there wasn't really a good way to get chrome/common/extensions without pulling in all of chrome/common. I'm not seeing any issues with the overlap in chromium_src/chrome/common and I know some of those files make electron-specific changes, but it seems like some of them are basically just copied over. Am I missing anything or can a lot of them be eliminated and added to `OTHER_HEADERS` in libchromiumcontent with the inclusion of libcommon?

**Comment by zcbenz** [comment]
@bridiver The source files of Chromium are copied to `chromium_src` when they:
a) need to be modified to be able to build with Electron;
b) are more easy to be maintained when copied over.

and I think most of the files have been modified, because the original design usually does not fit Electron API's usages.

I'm not sure whether building with everything under `chrome/common` would be a good idea for Electron, it is definitely necessary for a browser product, but probably more of a maintenance burden for Electron since we have large overlap on APIs. For fixing the devtools extension problem only, I'm more willing to add a simple JavaScript wrapper for `chrome.*` APIs.

**Comment by bridiver** [comment]
@zcbenz once I got it all working I was able to go back and remove the dependency on chrome/common (and just add a few more files to chromium_src).

**Comment by bridiver** [comment]
support for some other extensions took priority over getting things ready for upstream this week for me, but we need to rebase anyway to catch up so I plan on having a PR ready next week. Originally I tried to pull in as little as possible, but that actually ended up making things more difficult and complicated. They did a reasonably good job of separating extensions from chrome/\* and from the rest of the code, but it was still a bit of a tangled web. As with much of electron, I used the chrome shell as a template and tried to do as much as possible in JS, but a lot of it was less practical to do in JS than it seemed on the surface.
